ddtrace/internal/_encoding.pyx

```diff
@@ -106,6 +108,13 @@ cdef class Packer(object):
         cdef long i
 
         while True:
+            # Give other threads a chance to run every 16K "cycles"
+            if self.ticker & (16 << 10):
+                self.ticker = 0
+                with nogil: pass  
+            
+            self.ticker += 1
```

Comment:
This sounds like a very precise optimization, do you have serious data to back such an optimization? What's the impact before/after? 🤔

Reply:
Understood.

I'm still not a fan of such early optimizations trying to paper over CPython issues, especially without proper real-life evidence to back it up. It's not even clear if it's the right optimization we want.
I understand the code might not look like "much" here but that's a slippery slope of maintaining more code than necessary that I don't want to take.

Reply:
> I understand the code might not look like "much" here but that's a slippery slope of maintaining more code than necessary that I don't want to take.

Totally agree, but as I tried to stress out in the PR description, this is not an optimisation, but a refactor that can accommodate for other encoders with size constraint more easily. The ticker is there to have more control over the GIL. One of the benefits of doing partial encoding (as it's currently done) is the guarantee that all threads would get a chance to run while traces are being encoded. The ticker idea in this change makes that promise more reliable by ensuring that threads won't block for too long while a large trace is being encoded. Therefore, there shouldn't be any performance impacts coming from this change (just slightly better response time percentiles at most).

Reply:
In fact, after the re-refactor, the ticker change is somewhat unrelated to the main contents of the PR and could be taken out. What do folks think?

Reply:
I forgot to mention that this PR effectively **reduces** the amount of code to maintain while keeping the same functionality as before (modulo the ticker).